### PR TITLE
Blog post: Team Construction Pitfalls

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Adds a blog post talking about mistakes I've seen made in team construction. IMO a product team should have a balance of junior->senior and fresh->area-expert devs, and think hard about the time constraints it puts on itself.

- [x] Initial draft
- [x] Proofreading
- [x] Images
- [x] More proofreading
- [x] Final draft

### CODE REVIEW

File: content/blog/team-isolation-pitfalls/index.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+**TODO: IMAGE**
+
+<em style="display:block;margin-bottom:2rem;text-align:center;">
+“I know that I know nothing.” - Socrates, paraphrased by Plato, paraphrased
+<br />
+</em>
+
+Why is it that fresh new teams seem to always create tech debt before their first release?
+
+I'm not talking about the normal growing pains of expanding a small area of code; I mean the really divergent patterns that make existing developers think _"this is so off; why didn't they talk to us?"_.
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:16]: Should probably better specify this. It's not clear.

File: content/blog/team-isolation-pitfalls/index.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+## Preamble: Junior verses Senior
+
+**TODO: IMAGE**
+
+When does someone become a junior developer after learning to code?
+When do they become mid-range, or senior, or staff, or principle, or whatever else?
+What counts as "experience", and how do you measure it?
+
+I don't fully understand the answers to any of those questions.
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:35]: Maybe reword... my intent is to show these questions as nigh unsolveable, and something I have not yet fully understood myself (despite trying to)

File: content/blog/team-isolation-pitfalls/index.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+## 1. Overstocking Junior Developers
+
+**TODO: IMAGE**
+
+> Anecdote: an intern and a junior developer on a relatively new team were tasked with creating a new variation of an existing, tech-debt-heavy page.
+> They gave an initial estimate of about a week, started working on code, and quickly realized they'd need help.
+> They met with an experienced developer who gave advice on splitting components and code reuse.
+> Under perceived pressure to complete the asignment, however, they copy&pasted several folders of legacy code and edited them for the new page variant. 
+
+Some requirements of developer project planned are inherently difficult when you haven't already planned multiple projects.
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:48]: typo

File: content/blog/team-isolation-pitfalls/index.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+> The team opted to start product work quickly instead of writing RFCs or laying groundwork with platform owners.
+> Additions to existing code were made quickly and added an above expected amount of tech debt, the cleanups for which were added to the team backlog but not started on before launch.
+
+At what point does adding early tech debt for the sake of speed no longer worth it?
+Tech debt can be costly immediately when added to shared code areas, not just as an afterthought following a project's first launch.
+I have heard many justifications for the fast pace: proving a team's value; acquiring early user experience data to inform later experiments; critical business needs...
+These are all theoretically valid, but in practice I have developed a deep suspicion towards the frequency with which they're used.
+I've been burned too many times by a _perceived_ urgent need for release followed by long periods of inactivity.
+
+TODO: THIS IS NOT WELL SUPPORTED
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:106]: Wondering how to phrase this succinctly. I'm trying to say that the balance of considerations too often skews towards quick gains and doesn't consider tech debt that harms the team much sooner than expected, for a net negative within a quarter...

File: content/blog/team-isolation-pitfalls/index.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+---
+
+**TODO: IMAGE**
+
+<em style="display:block;margin-bottom:2rem;text-align:center;">
+“I know that I know nothing.” - Socrates, paraphrased by Plato, paraphrased
+<br />
+</em>
+
+Why is it that fresh new teams seem to always create tech debt before their first release?
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:14]: Maybe be more clear that this post is focusing on product teams? And maybe add a disclaimer somewhere about nothing being absolute?

File: content/blog/team-isolation-pitfalls/index.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+> The contractors were given a large, discrete chunk of the project.
+> After their code started to land into land into the main branch, other teams noticed it didn't use the existing design system components or recommended logical patterns.
+> Eventually the project was scrapped and code deleted for being too difficult to integrate with existing areas.
+
+There is no such thing as a ["Greenfield project"](https://en.wikipedia.org/wiki/Greenfield_project) in software development.
+In an existing organization, there should be established precedences and guidelines around how to structure code.
+Even in a brand new organization there are existing areas of familiarity for developers that should inform what to use.
+
+When an organization adds new developers, it is unhealthy to ask them to create code without onboarding them onto the existing guidelines and ongoing pushes of the team.
+To truly onboard an engineer onto a team's code takes months or more of having them work in the existing codebases with the existing engineers.
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:84]: > engineer

no, developer!

File: content/blog/team-isolation-pitfalls/index.md (hunk 5 above)
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:14]: Maybe note that this could also apply to design, etc.?

File: content/blog/team-isolation-pitfalls/index.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+Thus, it is unhealthy to spin off a new team solely with new team members, regardless of their general experience level.
+
+When you match area experts with developers new to the organization, both sides benefit.
+Area experts can inform the new developers on existing areas of investment and/or tech debt that should be factored into their work, along with why structures were set up that way.
+The new developers can bring in new proposals from the outside and work with the area experts to introduce new ideas to the team.
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:89]: Awkward phrasing.

File: content/blog/team-isolation-pitfalls/index.md (hunk 3 above)
Comment (johnnyreilly) [reply content/blog/team-isolation-pitfalls/index.md:48]: Haha - I came here to add that and you already have! Excellent!

File: content/blog/team-isolation-pitfalls/index.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+What an effective way to burn out your new developers!
+At the _very least_, when tackling larger tasks, they should be paired from the start with a more experienced developer who can assure them the assignment should be treated as a safe learning exercise.
+
+### Understocking Junior Developers
+
+Similarly, it is unhealthy to have too many _senior_ developers on product teams.
+While some teams -such as platform owners- might need many years of experience in each developer, most product teams are excellent opportunities to give your senior developers experience in mentoring juniors.
+Just as you don't want juniors to be impeded on their own, you also don't want seniors to stagnate in their knowledge sharing, intaking fresh new ideas, and ability to create understandable architectures.
+
+Junior developers provide a fresh perspective to a team that should not be underestimated.
```
Comment (johnnyreilly) [inline content/blog/team-isolation-pitfalls/index.md:67]: Completely agree with this. I worked with a team that moved from being almost entirely senior developers to being much more of a balance. Besides everything else it very much changed the nature of the team. It became more learning / teaching / sharing focussed and healthier as a consequence.

File: content/blog/team-isolation-pitfalls/index.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+**TODO: IMAGE**
+
+> Anecdote: a new team worked with upper management to propose what could be created in a year given company priorities.
+> Management created business goals and OKRs based on a very optimistic version of that timeline, referring to them as "aggressive goals" internally.
+> The team opted to start product work quickly instead of writing RFCs or laying groundwork with platform owners.
+> Additions to existing code were made quickly and added an above expected amount of tech debt, the cleanups for which were added to the team backlog but not started on before launch.
+
+At what point does adding early tech debt for the sake of speed no longer worth it?
+Tech debt can be costly immediately when added to shared code areas, not just as an afterthought following a project's first launch.
```
Comment (johnnyreilly) [inline content/blog/team-isolation-pitfalls/index.md:101]: I find this section interesting.  I too have experienced what I might call "forced pace" and I think it's nuanced.  Whilst the engineer in me finds it a frustrating, having a goal (generally in the form of a launch date) is a useful way to bring a team together that might otherwise be drifting (personal experience leaking in here) and make something happen.  

It reveals what's possible and what's going to be problematic.  Taking on tech debt (ideally as knowingly as possible) in this period seems reasonable; *with a caveat that I'll come to*.  Forced pace can be a useful way to avoid letting the perfect be the enemy of the good.  That's in favour of it.

The other side of the coin is where there's an attempt to repeat "forced pace" again and again *without* having paid down the tech debt that was built up towards the goal.  Regrettably I've seen this consistently done and it's a struggle.   One that I don't have an amazing answer for beyond advocating to allow for slack and ensure tech debt is paid down.  Measuring exactly how much tech debt is being paid down is a useful measure here to surface up. Likewise developer happiness which tends to have an inverse relationship with tech debt.

File: content/blog/team-isolation-pitfalls/index.md (hunk 5 above)
Comment (JoshuaKGoldberg) [reply content/blog/team-isolation-pitfalls/index.md:14]: Oh and _"always"_ is definitely not the right term here 😄. The intent is to complain that this happens sometimes, when the things complained about in this post are mishandled. _Sometimes_.

File: content/blog/team-isolation-pitfalls/index.md (hunk 9 above)
Comment (JoshuaKGoldberg) [reply content/blog/team-isolation-pitfalls/index.md:101]: Strong agreement! Thanks for the helpful phrasing; I'd been struggling to come up with how to better discuss this.

File: content/blog/team-isolation-pitfalls/index.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+“I know that I know nothing.” - Socrates, paraphrased by Plato, paraphrased
+<br />
+</em>
+
+Why is it that fresh new teams seem to always create tech debt before their first release?
+
+I'm not talking about the normal growing pains of expanding a small area of code; I mean the really divergent patterns that make existing developers think _"this is so off; why didn't they talk to us?"_.
+Maybe this isn't an issue you've experienced, but for me, it's been a consistent pain point with new product teams I've been on or partnered with over the last few years.
+
+In this blog post, I will provide vague anecdotal observations rail against and  to support the following three tenants:
```
Comment (callafeucht) [inline content/blog/team-isolation-pitfalls/index.md:19]: I'm not sure I understand your phrasing here:

>vague anecdotal observations _**rail against**_ and  to support the following three tenants

Also, you have an extra space between _and_ and _to_.

File: content/blog/team-isolation-pitfalls/index.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+3. Don't lie to yourself about timelines
+
+> Each anecdote is a real occurence I observed or participated in, though anonymized and slightly obfuscated.
+
+## Preamble: Junior verses Senior
+
+**TODO: IMAGE**
+
+When does someone become a junior developer after learning to code?
+When do they become mid-range, or senior, or staff, or principle, or whatever else?
```
Comment (MattSchiller) [inline content/blog/team-isolation-pitfalls/index.md:32]: super small nit, but the role is spelled `principal`

File: content/blog/team-isolation-pitfalls/index.md (hunk 2 above)
Comment (MattSchiller) [reply content/blog/team-isolation-pitfalls/index.md:35]: Perhaps something like "There is no one-size-fits-all answer to these questions"?

File: content/blog/team-isolation-pitfalls/index.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+For the sake of this easier writing, let's consider "junior" to be people who haven't yet gained repeat experience effectively performing the tasks described below, and "senior" to be those who have.
+
+## 1. Overstocking Junior Developers
+
+**TODO: IMAGE**
+
+> Anecdote: an intern and a junior developer on a relatively new team were tasked with creating a new variation of an existing, tech-debt-heavy page.
+> They gave an initial estimate of about a week, started working on code, and quickly realized they'd need help.
+> They met with an experienced developer who gave advice on splitting components and code reuse.
+> Under perceived pressure to complete the asignment, however, they copy&pasted several folders of legacy code and edited them for the new page variant.
```
Comment (MattSchiller) [inline content/blog/team-isolation-pitfalls/index.md:46]: also typo here on `assignment`. Not sure if that's what the below comment is referencing, but figured I'd drop a comment.

Loving this read!

File: content/blog/team-isolation-pitfalls/index.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+A healthy development team should pair its junior and senior developers together on advanced tasks such as project planning and architecture changes.
+Only a fool would "blame" the intern or junior developer from the anecdote.
+
+I'm not saying junior developers shouldn't participate in or even help lead major new investments -- practical experience is a great way to build seniority.
+Nor am I saying junior developers _can't_ build big things -- they can and should.
+I'm saying they ideally shouldn't do it _on their own_.
+
+Time and time again I've seen junior developers asked to do advanced tasks without adequate support, struggle to output as effectively as a senior, then blame themselves for reduced output.
+What an effective way to burn out your new developers!
+At the _very least_, when tackling larger tasks, they should be paired from the start with a more experienced developer who can assure them the assignment should be treated as a safe learning exercise.
```
Comment (MattSchiller) [inline content/blog/team-isolation-pitfalls/index.md:59]: grammar nit: Should it read `who can assure them (that) the assignment...`?

File: content/blog/team-isolation-pitfalls/index.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+Junior developers provide a fresh perspective to a team that should not be underestimated.
+If you don't think you need them, or if you don't think your teams have time for them, I question you sharply.
+
+## 2. Understocking Area Experts
+
+**TODO: IMAGE**
+
+> Anecdote: an existing team onboarded a set of contractors to help them create a more advanced version of their existing pages.
+> The contractors were given a large, discrete chunk of the project.
+> After their code started to land into land into the main branch, other teams noticed it didn't use the existing design system components or recommended logical patterns.
```
Comment (MattSchiller) [inline content/blog/team-isolation-pitfalls/index.md:76]: typo: `land into` x2

File: content/blog/team-isolation-pitfalls/index.md (hunk 9 above)
Comment (MattSchiller) [reply content/blog/team-isolation-pitfalls/index.md:101]: I might also add to this section that tech debt, like real debt, compounds on itself. So frequent, smaller pay-downs can be more effective than less periodic 'bulk' efforts

File: content/blog/team-isolation-pitfalls/index.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+**TODO: IMAGE**
+
+Mistakes in organizing large groups of people, developers or otherwise, are inevitable.
+I hope you can be well equipped to predict and mitigate some common pitfalls after reading this blog post:
+
+1. Don't deprive junior developers of senior developers _(or vice versa)_
+2. Don't deprive new developers of area experts
+3. Don't lie to yourself about timelines
+
+In conclusion: if you're setting up a product team, make sure it have a healthy mix of junior and senior developers, a healthy mix of fresh faces and area experts, and realistic timelines to strive for.
```
Comment (MattSchiller) [inline content/blog/team-isolation-pitfalls/index.md:128]: Loved this article. Thank you for putting this all together!

Comment (JoshuaKGoldberg) [reply content/blog/team-isolation-pitfalls/index.md:128]: Thanks for proofreading!

File: content/blog/team-isolation-pitfalls/index.md (hunk 10 above)
Comment (JoshuaKGoldberg) [reply content/blog/team-isolation-pitfalls/index.md:19]: ha, definitely some missing commas and added words here...

File: content/blog/team-isolation-pitfalls/index.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+I'm not talking about the normal growing pains of expanding a small area of code; I mean the really divergent patterns that make existing developers think _"this is so off; why didn't they talk to us?"_.
+Maybe this isn't an issue you've experienced, but for me, it's been a consistent pain point with new product teams I've been on or partnered with over the last few years.
+
+In this blog post, I will provide vague anecdotal observations rail against and  to support the following three tenants:
+
+1. Don't deprive junior developers of senior developers _(or vice versa)_
+2. Don't deprive new developers of area experts
+3. Don't lie to yourself about timelines
+
+> Each anecdote is a real occurence I observed or participated in, though anonymized and slightly obfuscated.
```
Comment (JoshuaKGoldberg) [inline content/blog/team-isolation-pitfalls/index.md:25]: I should probably also make it clear here that the obfuscation includes combining similar anecdotes over the >5 years.

File: content/blog/team-isolation-pitfalls/index.md (hunk 5 above)
Comment (JoshuaKGoldberg) [reply content/blog/team-isolation-pitfalls/index.md]: Meh, I'll stick to what I know. 😄 .

File: content/blog/team-isolation-pitfalls/index.md (hunk 13 above)
Comment (JoshuaKGoldberg) [reply content/blog/team-isolation-pitfalls/index.md]: I'm ok with it as is :)

File: content/blog/team-construction-pitfalls/index.md [hunk 17]
```diff
@@ -138,5 +138,5 @@ I hope you can be well equipped to predict and mitigate some common pitfalls aft
 
 In conclusion: if you're setting up a product team, make sure it have a healthy mix of junior and senior developers, a healthy mix of fresh faces and area experts, and realistic timelines to strive for.
 
-> Many thanks to the excellent [Johnny Reilly](https://twitter.com/johnny_reilly) for proofreading and discussions in the creation of this blog post, as well as my esteemed Codecademy coworkers: in particular [Calla Feucht](https://twitter.com/callacalla_) and [Matt Schiller](https://github.com/MattSchiller).
+> Many thanks to the excellent [John Reilly](https://twitter.com/johnny_reilly) for proofreading and discussions in the creation of this blog post, as well as my esteemed Codecademy coworkers: in particular [Calla Feucht](https://twitter.com/callacalla_) and [Matt Schiller](https://github.com/MattSchiller).
```
Comment (johnnyreilly) [inline content/blog/team-construction-pitfalls/index.md:141]: Haha! Yeah it's John. But way back when I was getting my first email address I found John was already taken. So I went with Johnny and I've used it as my handle pretty much everywhere ever since.

Comment (JoshuaKGoldberg) [reply content/blog/team-construction-pitfalls/index.md:141]: Haha makes sense; that's the same reason I put "K" in all my handles. No need to share an exact name with the attorney in NYC. Or the domestic terrorist from Florida...

Comment (johnnyreilly) [reply content/blog/team-construction-pitfalls/index.md:141]: Haha! I have my own version of that; there's already a semi famous John Reilly... But it's not me 😅

https://en.m.wikipedia.org/wiki/John_C._Reilly
